hw/mcu/nordic/nrf52xxx/src/hal_i2c.c

```diff
@@ -171,13 +192,13 @@ hal_i2c_clear_bus(const struct nrf52_hal_i2c_cfg *cfg)
     hal_gpio_write(cfg->scl_pin, 1);
     hal_gpio_write(cfg->sda_pin, 1);
 
-    /* Still with input buffer connected, the direction is output */
-    NRF_GPIO->DIRSET = ((1<<cfg->scl_pin)|(1<<cfg->sda_pin));
+    NRF_GPIO->PIN_CNF[cfg->scl_pin] = NRF52_SCL_PIN_CONF_CLR;
+    NRF_GPIO->PIN_CNF[cfg->sda_pin] = NRF52_SDA_PIN_CONF_CLR;
```

Comment:
You are changing the GPIO config, so you might want to do the following again after finishing clocking bits:
```
    NRF_GPIO->PIN_CNF[cfg->scl_pin] = NRF52_SCL_PIN_CONF;
    NRF_GPIO->PIN_CNF[cfg->sda_pin] = NRF52_SDA_PIN_CONF;
```

Reply:
Thanks, when the clear_bus command is run from outside of the hal_i2c_init function changing the io back is important. <- fix added. 

I'm not sure I'm following you on the second issue. The hal_gpio_read_in function I added does resolve the port and pins: 

```
static int
hal_gpio_read_in(int pin)
{
    NRF_GPIO_Type *port;
    port = HAL_GPIO_PORT(pin);

    return (port->IN >> HAL_GPIO_INDEX(pin)) & 1UL;
}
```

If the restoring of the pins' config is the problem for the nrf52840, then we have the same problem in hal_i2c_init() after the hal_i2c_clear_bus(); line, right?  

```
...    
    hal_i2c_clear_bus(cfg);

    NRF_GPIO->PIN_CNF[cfg->scl_pin] = NRF52_SCL_PIN_CONF;
    NRF_GPIO->PIN_CNF[cfg->sda_pin] = NRF52_SDA_PIN_CONF;

    regs->PSELSCL = cfg->scl_pin;
...
```

Reply:
Sorry, I should have been more clear `hal_gpio_read_in()` is correct as the implementation takes into account ports and specifics pins belonging to the ports. What seems to be problematic is assigning PIN_CNF to a gpio directly without selecting a particular port.